feat: [part 03] DHIS2-9637 adds new button and unifies the urls | unifies the url under new

paschalidi opened the pull request:
And this is the last part of this chain. 

What we do here is basically put together the work of the last 2 PRs. Now you can visit the `new` or registration page from the button and all the functionality should be as expected.

paschalidi commented on src/core_modules/capture-core/components/Buttons/SimpleSplitButton.component.js:
```diff
@@ -21,7 +22,7 @@ const SimpleSplitButton = (props: Props) => {
                     {
                         dropDownItems.map(i => (
-                            <MenuItem label={i.text} value={i.text} onClick={i.onClick} />
+                            <MenuItem key={uuid} label={i.text} value={i.text} onClick={i.onClick} />
```
irrelevant to the overall scope of the PR. However I got it fixed sine this line would throw an error in the console.

paschalidi replied:
I also prefer the `uuid` package. If you also agree we could go with that from now on. :)

JoakimSM replied:
yes, I agree. At some point we will get rid of d2-utilizr I think.

paschalidi replied:
Yes think so too.

paschalidi commented on src/core_modules/capture-core/components/DataEntries/SingleEventRegistrationEntry/DataEntryWrapper/DataEntry/DataEntry.component.js:
```diff
@@ -579,16 +579,9 @@ class NewEventDataEntry extends Component<Props> {
-                        {this.getSavingText()}
-                    </div>
-                </div>
+                <InfoIconText
+                    text={this.getSavingText()}
+                />
```
![image](https://user-images.githubusercontent.com/4181674/102347376-9f9b3b80-3fa0-11eb-8163-ff2ba6374292.png)
updates the component we used in the event programs for uniformity across the app

paschalidi commented on src/core_modules/capture-core/components/DataEntry/DataEntry.container.js:
```diff
@@ -43,5 +43,5 @@ const mapDispatchToProps = (dispatch: ReduxDispatch) => ({
 export default compose(
     // $FlowFixMe
     connect(mapStateToProps, mapDispatchToProps, null),
-    withLoadingIndicator(() => ({ height: '350px' }), null, ({ dataEntryExists }) => (dataEntryExists)),
```
had to revert this chance because it would throw an error in the console.

paschalidi commented on src/core_modules/capture-core/components/DataEntry/withCancelButton.js:
```diff
@@ -21,10 +21,6 @@ const getCancelButton = (InnerComponent: React.ComponentType<any>, optionsFn?: ?
 
         getWrappedInstance = () => this.innerInstance;
 
-        getCancelButtonInstance = (cancelButtonInstance: ?any) => {
-            this.props.cancelButtonRef && this.props.cancelButtonRef(cancelButtonInstance);
-        }
-
```
I might not understanding the whole spectrum of how this is used. It looked to me it wasnt in use.

paschalidi commented on src/core_modules/capture-core/components/LockedSelector/LockedSelector.container.js:
```diff
@@ -49,11 +48,16 @@ const mapDispatchToProps = (
     onResetAllCategoryOptions: () => {
         dispatch(resetAllCategoryOptionsFromLockedSelector());
     },
-    onOpenNewEventPage: (selectedProgramId, selectedOrgUnitId) => {
```
removed these two `selectedProgramId, selectedOrgUnitId` since in the epics we get this information from the redux state. We have been already doing this way. Meaning I didnt refactor this bit. These values must have been forgotten to be removed after some refactoring

paschalidi commented on src/core_modules/capture-core/components/Pages/New/NewPage.actions.js:
```diff
@@ -2,12 +2,16 @@ import { actionCreator } from '../../../actions/actions.utils';
 export const showMessageToSelectOrgUnitOnNewPage = () =>
     actionCreator(newPageActionTypes.NEW_PAGE_WITHOUT_ORG_UNIT_SELECTED_VIEW)();
 
+export const showMessageToSelectProgramPartnerOnNewPage = () =>
+    actionCreator(newPageActionTypes.NEW_PAGE_WITHOUT_PROGRAM_PARTNER_SELECTED_VIEW)();
+
```
We use this to display message to the user when they have selected a program which has partners but they havent selected a partner.

Bekkalizer replied:
Yes, "Implementing partner" is the name for the category - I guess that is where the term "Partner" comes from, and we should avoid using this term as @JoakimSM points out. Other programs will have other categories, and the names of these categories is defined in the category metadata. I think for now we change this to "category" as the generic term.

paschalidi replied:
So the text when someone hasnt selected a category would be "Choose a category to start reporting".

paschalidi commented on src/core_modules/capture-core/components/Pages/New/NewPage.component.js:
```diff
@@ -2,123 +2,70 @@
-                        </div>
-                    </div>
-                </Paper>
+                <RegistrationDataEntry
+                    dataEntryId={NEW_TEI_DATA_ENTRY_ID}
+                    selectedScopeId={selectedScopeId}
+                    setScopeId={setScopeId}
+                />
```
I thought this makes the overall component a little bit easier to comprehend.

paschalidi commented on src/core_modules/capture-core/reducers/descriptions/activePage.reducerDescription.js:
```diff
@@ -10,10 +10,6 @@ import { actionTypes as viewEventDataEntryActionTypes } from '../../components/P
 import { eventWorkingListsActionTypes } from '../../components/Pages/MainPage/EventWorkingLists';
 
 export const activePageDesc = createReducerDescription({
-    [lockedSelectorActionTypes.ORG_UNIT_ID_SET]: state => ({
-        ...state,
-        isDataEntryLoading: true,
-    }),
```
also this caused some weird behaviour and didnt see where we make use of it. There is good chance that I am simply unaware of the use-case so please take a look.

paschalidi commented on src/core_modules/capture-core/reducers/descriptions/currentSelections.reducerDescriptionGetter.js:
```diff
@@ -47,21 +48,24 @@ const setCategoryOption = (
         ...state,
-        categories,
-        categoriesMeta,
+        categories: newCategories,
+        categoriesMeta: newCategoriesMeta,
```
Regarding my todo here. I was thinking, cant we simply have these two lines here be undefined?

What we do now is that each tome a user deselects a partner selection from the LockedSelector we remove that specific selection only from the object.

JoakimSM replied:
There can be multiple category selections for a program (partner is just a preconfigured category in the demo db) so we need to keep this logic.

paschalidi replied:
ah right now I get it. makes sense

JoakimSM commented on src/core_modules/capture-core/reducers/descriptions/currentSelections.reducerDescriptionGetter.js:
```diff
@@ -233,4 +237,12 @@ export const getCurrentSelectionsReducerDesc = (appUpdaters: Updaters) => create
+}, 'currentSelections', {
+    complete: false,
+    categoryCheckInProgress: false,
+    orgUnitId: null,
+    programId: null,
+    trackedEntityTypeId: null,
+    categories: null,
+    categoriesMeta: null,
```
I would be careful assigning null here, because this will override default function param assignments. Especially for categories and categoriesMeta which are saved as objects. I suggest assigning to undefined or not at all.

paschalidi replied:
very good point! thanks

JoakimSM reviewed (commented):
Good work @paschalidi. Made some comments, one regarding "partner" that we need to discuss.

JoakimSM commented:
Another point I wanted to make; earlier when you navigated away from the page without saving your changes you would get a message: 
![image](https://user-images.githubusercontent.com/12745737/103761159-85400500-5016-11eb-93d3-3eb43d2c55a0.png)
I don't know if you already have a fix for this in the pipeline.

paschalidi commented:
Right. I ve missed that. Thanks

paschalidi commented:
https://jira.dhis2.org/browse/DHIS2-10206 @JoakimSM I have create this ticket

paschalidi commented on src/core_modules/capture-core/components/Pages/New/RegistrationDataEntry/RegistrationDataEntry.component.js:
```diff
@@ -1,16 +1,20 @@
 import i18n from '@dhis2/d2-i18n';
-import type { OwnProps } from './RegistrationDataEntry.types';
-import { EnrollmentRegistrationEntry, TeiRegistrationEntry } from '../../../DataEntries';
+import Paper from '@material-ui/core/Paper/Paper';
```
makes sense

paschalidi commented:
@JoakimSM looked in your comments this now again ready for review

JoakimSM commented on src/core_modules/capture-core/components/Pages/New/NewPage.component.js:
```diff
@@ -84,10 +85,16 @@ const NewPagePlain = ({
+                    const { name = 'program category' } = categoryCombination || {};
+                    return (
+                        <InefficientSelectionsMessage
+                            message={i18n.t('Choose a {{name}} to start reporting', { name })}
```
A category combination can consist of mulitple categories. I think we should list the category names here instead of using the category combination name. We can discuss on slack or in our upcoming meeting.

paschalidi replied:
![image](https://user-images.githubusercontent.com/4181674/104012039-6c1e8c00-51af-11eb-8af7-4e654c03515c.png)
This looks like this but when for some reason `categoryCombination.name` is undefined then initial value of the name `program category`.

JoakimSM commented:
> https://jira.dhis2.org/browse/DHIS2-10206 @JoakimSM I have create this ticket. Is it alright if we tackle this on a separate PR?

Certainly. Perfect.

JoakimSM commented on src/core_modules/capture-core/components/Pages/New/NewPage.component.js:
```diff
@@ -135,6 +83,24 @@ const NewPagePlain = ({
+                    }, '');
+
+                    return (
+                        <InefficientSelectionsMessage
```
NIT: Maybe Insufficient/Incomplete would be better?

paschalidi replied:
Yep good point. I am changing it

JoakimSM commented:
The category combo -> categories change looks good. Could you, as a last effort on this, remove everything referencing 'partner'? In this PR or a new PR, it's up to you.

paschalidi commented:
@JoakimSM I think I have changed everything that was referring to partner. Do you reckon I missed something? 
Also just renamed the `InefficientSelectionsMessage` to `IncompleteSelectionsMessage`.

JoakimSM commented:
> @JoakimSM I think I have changed everything that was referring to partner. Do you reckon I missed something?
> Also just renamed the `InefficientSelectionsMessage` to `IncompleteSelectionsMessage`.

For example `showMessageToSelectProgramPartnerOnNewPage`. I searched for partner and got 10 hits.

paschalidi commented:
Okay I was in the wrong branch :| I am applying the changes here and will poke you on slack for another review
